### PR TITLE
Partial evaluation support for branching

### PR DESCRIPTION
**Pull request by cesarzc**
This change implements support for branching in partial evaluation.

### CODE REVIEW

File: compiler/qsc_partial_eval/src/lib.rs [hunk 1]
```diff
@@ -519,6 +550,152 @@ impl<'a> PartialEvaluator<'a> {
...
+                let maybe_otherwise_value = self.try_eval_expr(otherwise_expr_id);
+                if let Ok(otherwise_value) = maybe_otherwise_value {
+                    Ok(otherwise_value)
+                } else {
+                    let otherwise_expr = self.get_expr(otherwise_expr_id);
+                    let error = Error::FailedToEvaluateIfExpressionBranchBlock(otherwise_expr.span);
+                    Err(error)
+                }
+            } else {
+                Ok(Value::unit())
```
Comment (cesarzc) [inline compiler/qsc_partial_eval/src/lib.rs:590]: Explain this case.

File: compiler/qsc_partial_eval/src/lib.rs [hunk 2]
```diff
@@ -519,6 +550,152 @@ impl<'a> PartialEvaluator<'a> {
...
+        let maybe_condition_value = self.try_eval_expr(condition_expr_id);
+        let Ok(condition_value) = maybe_condition_value else {
+            let condition_expr = self.get_expr(condition_expr_id);
+            let error = Error::FailedToEvaluateConditionExpression(condition_expr.span);
+            return Err(error);
+        };
+
+        // If the condition value is a Boolean literal, use the value to decide which branch to
+        // evaluate.
+        if let Value::Bool(condition_bool) = condition_value {
```
Comment (cesarzc) [inline compiler/qsc_partial_eval/src/lib.rs:570]: Encapsulate in function.

File: compiler/qsc_partial_eval/src/lib.rs [hunk 3]
```diff
@@ -519,6 +550,152 @@ impl<'a> PartialEvaluator<'a> {
...
+            let maybe_if_expr_value = if condition_bool {
+                let maybe_body_value = self.try_eval_expr(body_expr_id);
+                if let Ok(body_value) = maybe_body_value {
```
Comment (cesarzc) [inline compiler/qsc_partial_eval/src/lib.rs:573]: This can be made more compact.

File: compiler/qsc_partial_eval/src/lib.rs [hunk 4]
```diff
@@ -519,6 +550,152 @@ impl<'a> PartialEvaluator<'a> {
...
+
+        // Evaluate the body expression.
+        let (body_block_node_id, body_expr_value) =
+            self.eval_expr_if_branch(body_expr_id, continuation_block_node_id)?;
+        // If there is a variable to save to, add a store instruction at the end of the body block.
+        if let Some(if_expr_var) = maybe_if_expr_var {
+            let body_operand = map_eval_value_to_rir_operand(&body_expr_value);
+            let store_ins = Instruction::Store(body_operand, if_expr_var);
+            let body_block = self.get_program_block_mut(body_block_node_id);
+            body_block.0.push(store_ins);
```
Comment (cesarzc) [inline compiler/qsc_partial_eval/src/lib.rs:628]: This is a bug because this store instruction will happen after the jump instruction.

Comment (cesarzc) [reply compiler/qsc_partial_eval/src/lib.rs:628]: Add test that fails.

File: compiler/qsc_partial_eval/src/lib.rs [hunk 5]
```diff
@@ -519,6 +550,152 @@ impl<'a> PartialEvaluator<'a> {
...
+                let otherwise_block = self.get_program_block_mut(otherwise_block_node_id);
+                otherwise_block.0.push(store_ins);
+            }
+            otherwise_block_node_id
+        } else {
+            continuation_block_node_id
+        };
+
+        // Finally, we insert the branch instruction.
+        let condition_var = if let Value::Var(var) = condition_value {
```
Comment (cesarzc) [inline compiler/qsc_partial_eval/src/lib.rs:648]: should be `let else`.

**Review by idavis — APPROVED** [review]

File: compiler/qsc_partial_eval/src/lib.rs (hunk 5 above)
Comment (cesarzc) [reply compiler/qsc_partial_eval/src/lib.rs:648]: It actually needs to be an if-expression

**Pull request merged**